File: nbs_customization/controllers/item.py
import frappe
from frappe import _


def validate(doc, method):
	if doc.custom_instrument_specification and doc.custom_reagent_specification:
		frappe.throw(
			_("Item cannot be both an instrument and a reagent. "
			  "Set only one of Instrument Specification or Reagent Specification.")
		)

	if doc.custom_instrument_specification:
		spec_item = frappe.db.get_value(
			"Instrument Specification", doc.custom_instrument_specification, "item")
		if spec_item and spec_item != doc.name:
			frappe.throw(
				_("Instrument Specification {0} is for {1}, not {2}.").format(
					frappe.bold(doc.custom_instrument_specification),
					frappe.bold(spec_item),
					frappe.bold(doc.name),
				))

	if doc.custom_reagent_specification:
		spec_item = frappe.db.get_value(
			"Reagent Specification", doc.custom_reagent_specification, "item")
		if spec_item and spec_item != doc.name:
			frappe.throw(
				_("Reagent Specification {0} is for {1}, not {2}.").format(
					frappe.bold(doc.custom_reagent_specification),
					frappe.bold(spec_item),
					frappe.bold(doc.name),
				))

File: nbs_customization/public/js/item.js
frappe.ui.form.on("Item", {
	setup(frm) {
		_setup_placement_fields(frm);
	},

	custom_is_placement_item(frm) {
		_toggle_placement_fields(frm);
	},
});

function _setup_placement_fields(frm) {
	_toggle_placement_fields(frm);

	frm.set_query("custom_instrument_specification", function () {
		return {
			filters: {
				item: ["in", [frm.doc.name, ""]],
			},
		};
	});

	frm.set_query("custom_reagent_specification", function () {
		return {
			filters: {
				item: ["in", [frm.doc.name, ""]],
			},
		};
	});
}

function _toggle_placement_fields(frm) {
	const show = frm.doc.custom_is_placement_item === 1;

	if (show) {
		frm.set_df_property(
			"custom_instrument_specification",
			"description",
			__(
				"Select the Instrument Specification if this item is an analyzer. Defines supported tests and consumables.",
			),
		);
		frm.set_df_property(
			"custom_reagent_specification",
			"description",
			__("Select the Reagent Specification if this item is a reagent or consumable."),
		);
	}
}
